File: Baekjoon/그래프와 순회/MinjuCho/16928.py
# 문제 풀이 링크: https://velog.io/@cuppizza/백준-16928-뱀과-사다리-게임-C-파이썬-BFS-최단거리
# 실행 시간: 100ms 메모리: 36888KB
from queue import deque
import sys
input = sys.stdin.readline
n, m = map(int, input().strip().split())
ladders = {}
snakes = {}
for i in range(n):
    x, y = map(int, input().strip().split())
    ladders[x] = y
for i in range(m):
    u, v = map(int, input().strip().split())
    snakes[u] = v
ladders_keys = set(ladders.keys())
snakes_keys = set(snakes.keys())
graph = [0 for _ in range(101)]
# BFS
q = deque([1])
graph[1] = 1
while q:
    cur = q.popleft()
    for i in range(1, 7):
        if cur + i <= 100 and graph[cur + i] == 0:
            graph[cur + i] = graph[cur] + 1
            if cur + i in ladders_keys and graph[ladders[cur + i]] == 0:
                graph[ladders[cur + i]] = graph[cur + i]
                q.append(ladders[cur + i])
                continue
            elif cur + i in snakes_keys and graph[snakes[cur + i]] == 0:
                graph[snakes[cur + i]] = graph[cur + i]
                q.append(snakes[cur + i])
                continue
            elif (cur + i in ladders_keys and graph[ladders[cur + i]] != 0) \
                 or (cur + i in snakes_keys and graph[snakes[cur + i]] != 0):
                continue
            q.append(cur + i)
print(graph[100] - 1)

File: Baekjoon/그래프와 순회/MinjuCho/16928.cpp
// 문제 풀이 링크: https://velog.io/@cuppizza/백준-16928-뱀과-사다리-게임-C-파이썬-BFS-최단거리
// 실행 시간: 0ms 메모리: 2028KB
#include <map>
#include <queue>
#include <vector>
#include <iostream>

using namespace std;
int main(void)
{
    int n, m;
    cin >> n >> m;
    map<int, int> ladders, snakes;
    int x, y;
    for (int i=0; i<n; ++i)
    {
        cin >> x >> y;
        ladders[x] = y;
    }
    for (int i=0; i<m; ++i)
    {
        cin >> x >> y;
        snakes[x] = y;
    }
    queue<int> q;
    q.push(1);
    vector<int> board(101, 0);
    board[1] = 1;
    int cur;
    while(!q.empty())
    {
        cur = q.front();
        q.pop();
        for(int i=1; i<=6; ++i)
        {
            if (cur+i <= 100 && board[cur+i] == 0)
            {
                board[cur+i] = board[cur] + 1;
                if(ladders[cur+i] != 0 && board[ladders[cur+i]] == 0)
                {
                    board[ladders[cur+i]] = board[cur+i];
                    q.push(ladders[cur+i]);
                    continue;
                }
                else if(snakes[cur+i] != 0 && board[snakes[cur+i]] == 0)
                {
                    board[snakes[cur+i]] = board[cur+i];
                    q.push(snakes[cur+i]);
                    continue;
                }
                else if ((snakes[cur+i] != 0 && board[snakes[cur+i]] != 0)||(ladders[cur+i] != 0 && board[ladders[cur+i]] != 0))
                {
                    continue;
                }
                q.push(cur+i);
            }
        }
    }
    cout << board[100]-1;
}